### ISSUE TITLE
Hide/mask/scrub secrets/passwords in logs

### ISSUE THREAD

**Issue by ohadschn** [appium/appium#7622]
## The problem
Passwords and other secrets sent by `sendKeys` variants are seen in plaintext in the logs. It would be nice to be able to prevent that somehow.

## Details

Suppose I have code that looks like this
```js
var pass = getPassworSecurelySomehow(); //e.g. DPAPI or keychain
myPasswordElement.sendKeys(pass);
```

The logs will show my password:

> [debug] Pushing command to appium work queue: "au.getElement('18').setValueByType('MySecretPass)"

It would be great to be able to write something like ```myPasswordElement.sendSecureKeys(pass);``` (without compromising log verbosity), and have those keys scrubbed from the log (e.g. replaced with asterisks).

I realize that this would require client library support too (e.g. in `wd`) but I figured the support has to start somewhere, and could possibly even exploit some existing mechanism (maybe send a special character value that would mark the text as 'secret')?

**Comment by dpgraham** [comment]
What is the use-case you were envisioning? Would it be when the testers use their own password in an Appium test and then when a 2nd or 3rd party looks at the test logs you don't want them to be able to see the password?

@imurchie @jlipps Do we have a way to scrub PII from server logs?

**Comment by imurchie** [comment]
We don't. And it is hard to envision a way to do it that doesn't have unwanted side-effects.

**Comment by ohadschn** [comment]
@dpgraham yes something like that, it doesn't have to be your password, any password really shouldn't be visible in plaintext. Compromised "test" accounts are often more dangerous than "real" accounts, for example they usually have 2FA disabled and monitored less closely.

@imurchie could you elaborate please? Forgive me as I know nothing about appium source code..

**Comment by imurchie** [comment]
Making a new client-side method and a new endpoint on the server would work. I was thinking of the "special character" option, which almost never works, since what is a special character is a normal one to some use-cases.

I'll add this to the backlog to try to address.

**Comment by jlipps** [comment]
yeah, maybe adding a custom json parameter `redact` to the sendkeys method would work. We can bring this up with the selenium folks too to see if it's worth talking about in terms of the spec.

@lukeis @shs96c have you guys had requests before for a flag for sendkeys that will instruct the server not to log the text in case it's sensitive information?

Alternatively, we could stay within spec and simply have a mode for the Appium server where it doesn't log _any_ sendkeys information

**Comment by ohadschn** [comment]
@jlipps 

> simply have a mode for the Appium server where it doesn't log any sendkeys information

That's a great idea, at the very least could serve as a stopgap until the redact approach is implemented.

**Comment by dotnetprofessional** [comment]
Seems this hasn't had much traction. We've just run into this issue ourselves. As @ohadschn has mentioned these 'test' passwords can be more dangerous than the real ones. They are often used to gain access to other systems even if the original password doesn't provide access. Many organizations have a policy of not logging passwords or committing the to source control (another popular source for hackers).

I like the idea of an optional parameter to the sendKeys method that prevents that data from being logged. I'd go with isSensitive, but redact is fine too.

Is this likely to be added?

**Comment by arushi-mac** [comment]
I can still see passwords in the logs, not sure what happened with this issue. The ticket is marked as closed, it would be nice if someone can provide an update. Thanks!

**Comment by mykola-mokhnach** [comment]
https://github.com/appium/appium/pull/14239 does implement the desired functionality. It will be available as soon as the next appium beta version is published.

**Comment by mykola-mokhnach** [comment]
The feature is available since 1.18.0-beta
